Services/Mail/classes/Autoresponder/AutoresponderServiceImpl.php

```diff
@@ -0,0 +1,151 @@
+
+        $now = $this->normalizeDateTimezone($this->clock->now());
+
+        // Don't compare the objects because of microseconds
+        return $last_send_time_with_added_interval->format('Y-m-d H:i:s') <= $now->format('Y-m-d H:i:s');
+    }
+
+    public function isAutoresponderEnabled(): bool
+    {
+        return $this->auto_responder_status;
+    }
+
+    public function enableAutoresponder(): void
+    {
+        $this->auto_responder_status = true;
+    }
+
+    public function disableAutoresponder(): void
+    {
+        $this->auto_responder_status = false;
+    }
+
+    public function handleAutoresponderMails(int $auto_responder_receiver_usr_id): void
+    {
+        if ($this->auto_responder_data === []) {
+            return;
+        }
+
+        foreach ($this->auto_responder_data as $auto_responder_sender_usr_id => $mail_options) {
+            if ($this->auto_responder_repository->exists($auto_responder_sender_usr_id, $auto_responder_receiver_usr_id)) {
+                $auto_responder = $this->auto_responder_repository->findBySenderIdAndReceiverId(
+                    $auto_responder_sender_usr_id,
+                    $auto_responder_receiver_usr_id
+                );
+            } else {
+                $auto_responder = new AutoresponderDto(
+                    $auto_responder_sender_usr_id,
+                    $auto_responder_receiver_usr_id,
+                    $this->normalizeDateTimezone($this->clock->now())
+                         ->sub($this->idle_time_interval)
+                         ->modify('-1 second')
+                );
+            }
+
+            if ($this->shouldSendAutoresponder($auto_responder)) {
+                $auto_responder = $auto_responder->withSentTime($this->clock->now());
+
+                if ($this->mail_action !== null) {
+                    ($this->mail_action)(
+                        $auto_responder_sender_usr_id,
+                        $mail_options,
+                        $auto_responder->getSentTime()->add($this->idle_time_interval)
+                    );
+                } else {
+                    $mail = new AutoresponderNotification(
+                        $auto_responder_sender_usr_id,
+                        $mail_options,
+                        $auto_responder->getSentTime()->add($this->idle_time_interval)
+                    );
+                    $mail->send();
+                }
+
+                $this->auto_responder_repository->store($auto_responder);
+            }
+        }
+    }
+
+    public function enqueueAutoresponderIfEnabled(ilMailOptions $mail_recipient_mail_options, int $receiver_id): void
+    {
+        if ($this->auto_responder_status && $mail_recipient_mail_options->isAbsent()) {
+            $sender_id = $mail_recipient_mail_options->getUsrId();
+            $mail_recipient_mail_options = clone $mail_recipient_mail_options;
+            $reflection_usr_id = new \ReflectionProperty(ilMailOptions::class, 'usrId');
```

Comment:
@mjansenDatabay Is there a better way for the autoresponder mail back to sender?

Reply:
Hmmm, there is (almost) always a better approach then using introspection, at least if 1 is maintainer of the affected code.

Does it help to add a new public method `getUserId`?

Reply:
https://github.com/ILIAS-eLearning/ILIAS/pull/4885/files#diff-6347dce422a6e5cf515d769bee07abc81edca22ff56a565019e9bd058f4f86c8R732

i changed the passed mail_options, this should do the trick